README
= Tinder

Tinder is a library for interfacing with Campfire, the chat application from 37Signals.

  campfire = Campfire.new 'mysubdomain'
  campfire.login 'myemail@example.com', 'mypassword'
  room = campfire.create_room 'New Room', 'My new campfire room to test tinder'
  room.rename 'New Room Name'
  room.speak 'Hello world!'
  room.speak "my pasted\ncode", :paste => true
  room.destroy

== Requirements

* Hpricot
  gem install hpricot

== ToDo

* Tests!
* Log in via guest url

CHANGELOG.txt
1.10.1 - 2014-12-01
* Loosen hashie dependency

1.10.0 - 2014-08-06
* Update Faraday dependency
* Loosen mime-type dependency

1.9.4 - 2014-02-24
* Fix typos in MIT licence
* Add Campfire#presence to get a list of rooms in which you are present

1.9.3 - 2013-09-06
* OAuth support

1.9.2 - 2013-01-04
* Update dependencies to latest versions

1.9.0 - 2012-07-16
* Add Room#recent to get a list of recent messages
* Add Room#search to search a room's transcripts

1.4.3 - 2010-12-07
* explicitly require 'uri'
* added Room#tweet(url)

1.4.2 - 2010-11-13
* Use Faraday instead of HTTParty [eric]
* Fix file uploads [eric]

1.4.1 - 2010-10-09
* Make SSL the default since it is available for all Campfire accounts.
* Added MIT License

1.4 - 2010-05-11
* Remove methods no longer supported by API
  Campfire#available_transcripts, Room#ping, Room#destroy, Room#toggle_guest_access
* Added Room#play
* ActiveSupport 3.0 support
* Fix streaming API support
* Allow SSL for listening
* Add support for HTTP proxies [c13bcc0b]

1.3.1 - 2009-12-17
* Declare HTTParty dependency
* Fix Room#paste

1.3.0 - 2009-12-15
* Rewrite to use Official Campfire API

1.2.2 - 2009-09-12
* Work around CSRF protection bug
* Fixes for changes to Campfire markup
* Include timestamps in the transcript

1.2.1 - 2009-08-27
* Fixes for listening after campfire updates [Jordan Byron]

1.2.0 - 2009-01-28
* Get the list of available files [Christopher MacGown]
* Upload files [Joshua Wand]
* Find rooms even when full [Josh Owens]
* Join rooms as a guest [Ian Lesperance]

1.1.7 - 2008-07-24
* Don't join the room when only speaking [Brian Donovan]
* Added support for HTTP proxies
* Fix listening for messages that contain URLs [Jared Kuolt]

0.1.6 - 2008-03-07
* Added Room#topic for getting the current topic [Even Weaver]
* Trap INT in #listen(&block) [borrowed from Chris Shea's Pyre]

0.1.5 - 2008-01-25
* Fixed Room#listen, which was broken by latest Campfire deploy
* Fixed timeout when listening but not speaking that will eventually log you out [Clinton R. Nixon]

0.1.4 - 2007-07-23
* Support for transcripts
* Fixed Room#leave, which was broken by a Campfire deployment [Andy Smith]

0.1.3 - 2007-02-12
* added ssl support [Tero Parviainen]

0.1.2 - 2007-01-27
* fixed bug preventing #listen from working without a block

0.1.1 - 2007-01-27
* fix bug preventing speak from working
* incorporated "watching" from http://soylentfoo.jnewland.com/articles/2006/12/07/updates-to-marshmallow-the-campfire-bot

0.1.0 - 2007-01-23
* Initial release as gem
* Get the users in a room [Tero Parviainen]


lib/tinder/room.rb
# encoding: UTF-8
require 'time'

module Tinder
  # A campfire room
  class Room
    attr_reader :id, :name

    def initialize(connection, attributes = {})
      @connection = connection
      @id = attributes['id']
      @name = attributes['name']
      @loaded = false
    end

    # Join the room
    # POST /room/#{id}/join.xml
    # For whatever reason, #join() and #leave() are still xml endpoints
    # whereas elsewhere in this API we're assuming json :\
    def join
      post 'join', 'xml'
    end

    # Leave a room
    # POST /room/#{id}/leave.xml
    def leave
      post 'leave', 'xml'
      stop_listening
    end

    # Get the url for guest access
    def guest_url
      "#{@connection.uri}/#{guest_invite_code}" if guest_access_enabled?
    end

    def guest_access_enabled?
      load
      @open_to_guests ? true : false
    end

    # The invite code use for guest
    def guest_invite_code
      load
      @active_token_value
    end

    # Change the name of the room
    def name=(name)
      update :name => name
    end
    alias_method :rename, :name=

    # Change the topic
    def topic=(topic)
      update :topic => topic
    end

    def update(attrs)
      connection.put("/room/#{@id}.json", {:room => attrs})
    end

    # Get the current topic
    def topic
      reload!
      @topic
    end

    # Lock the room to prevent new users from entering and to disable logging
    def lock
      post 'lock'
    end

    # Unlock the room
    def unlock
      post 'unlock'
    end

    # Post a new message to the chat room
    def speak(message, options = {})
      send_message(message)
    end

    def paste(message)
      send_message(message, 'PasteMessage')
    end

    def play(sound)
      send_message(sound, 'SoundMessage')
    end

    def tweet(url)
      send_message(url, 'TweetMessage')
    end

    # Get the list of users currently chatting for this room
    def users
      @users ||= current_users
    end

    def current_users
      reload!
      @current_users
    end

    # return the user with the given id; if it isn't in our room cache,
    # do a request to get it
    def user(id)
      if id
        cached_user = users.detect {|u| u[:id] == id }
        user = cached_user || fetch_user(id)
        self.users << user
        user
      end
    end

    # Perform a request for the user with the given ID
    def fetch_user(id)
      user_data = connection.get("/users/#{id}.json")
      user = user_data && user_data[:user]
      user[:created_at] = Time.parse(user[:created_at])
      user
    end

    # Modifies a hash representation of a Campfire message.  Expands +:user_id+
    # to a full hash at +:user+, generates Timestamp from +:created_at+.
    #
    # Full returned hash:
    # * +:body+: the body of the message
    # * +:user+: Campfire user, which is itself a hash, of:
    #   * +:id+: User id
    #   * +:name+: User name
    #   * +:email_address+: Email address
    #   * +:admin+: Boolean admin flag
    #   * +:created_at+: User creation timestamp
    #   * +:type+: User type (e.g. Member)
    # * +:id+: Campfire message id
    # * +:type+: Campfire message type
    # * +:room_id+: Campfire room id
    # * +:created_at+: Message creation timestamp
    def parse_message(message)
      message[:user] = user(message.delete(:user_id))
      message[:created_at] = Time.parse(message[:created_at])
      message
    end

    # Listen for new messages in the room, parsing them with #parse_message
    # and then yielding them to the provided block as they arrive.
    #
    #   room.listen do |m|
    #     room.speak "Go away!" if m[:body] =~ /Java/i
    #   end
    def listen(options = {})
      raise ArgumentError, "no block provided" unless block_given?

      Tinder.logger.info "Joining #{@name}…"
      join # you have to be in the room to listen

      require 'json'
      require 'hashie'
      require 'multi_json'
      require 'twitter/json_stream'

      auth = connection.basic_auth_settings
      options = {
        :host => "streaming.#{Connection::HOST}",
        :path => room_url_for('live'),
        :auth => "#{auth[:username]}:#{auth[:password]}",
        :timeout => 6,
        :ssl => connection.options[:ssl]
      }.merge(options)

      Tinder.logger.info "Starting EventMachine server…"
      EventMachine::run do
        @stream = Twitter::JSONStream.connect(options)
        Tinder.logger.info "Listening to #{@name}…"
        @stream.each_item do |message|
          message = Hashie::Mash.new(MultiJson.decode(message))
          message = parse_message(message)
          yield(message)
        end

        @stream.on_error do |message|
          raise ListenFailed.new("got an error! #{message.inspect}!")
        end

        @stream.on_max_reconnects do |timeout, retries|
          raise ListenFailed.new("Tried #{retries} times to connect. Got disconnected from #{@name}!")
        end

        # if we really get disconnected
        raise ListenFailed.new("got disconnected from #{@name}!") if !EventMachine.reactor_running?
      end
    end

    def listening?
      @stream != nil
    end

    def stop_listening
      return unless listening?

      Tinder.logger.info "Stopped listening to #{@name}…"
      @stream.stop
      @stream = nil
    end

    # Get the transcript for the given date (Returns a hash in the same format as #listen)
    #
    #   room.transcript(Time.now)
    #   #=> [{:message=>"foobar!",
    #         :user_id=>"99999",
    #         :person=>"Brandon",
    #         :id=>"18659245",
    #         :timestamp=>=>Tue May 05 07:15:00 -0700 2009}]
    #
    # The timestamp slot will typically have a granularity of five minutes.
    #
    def transcript(transcript_date = Date.today)
      url = "/room/#{@id}/transcript/#{transcript_date.strftime('%Y/%m/%d')}.json"
      connection.get(url)['messages'].map do |message|
        parse_message(message)
      end
    end

    # Search transcripts for the given term (returns an array of messages parsed
    # via #parse_message, see #parse_message for format of returned message)
    #
    def search(term)
      encoded_term = URI.encode(term)

      room_messages = connection.get("/search/#{encoded_term}.json")["messages"].select do |message|
        message[:room_id] == id
      end

      room_messages.map do |message|
        parse_message(message)
      end
    end

    def upload(file, content_type = nil, filename = nil)
      require 'mime/types'
      content_type ||= MIME::Types.type_for(filename || file)
      raw_post(:uploads, { :upload => Faraday::UploadIO.new(file, content_type, filename) })
    end

    # Get the list of latest files for this room
    def files(count = 5)
      get(:uploads)['uploads'].map { |u| u['full_url'] }
    end

    # Get a list of recent messages
    # Accepts a hash for options:
    # * +:limit+: Restrict the number of messages returned
    # * +:since_message_id+: Get messages created after the specified message id
    def recent(options = {})
      options = { :limit => 10, :since_message_id => nil }.merge(options)
      # Build url manually, faraday has to be 8.0 to do this
      url = "#{room_url_for(:recent)}?limit=#{options[:limit]}&since_message_id=#{options[:since_message_id]}"

      connection.get(url)['messages'].map do |msg|
        parse_message(msg)
      end
    end

  protected

    def load
      reload! unless @loaded
    end

    def reload!
      attributes = connection.get("/room/#{@id}.json")['room']

      @id = attributes['id']
      @name = attributes['name']
      @topic = attributes['topic']
      @full = attributes['full']
      @open_to_guests = attributes['open_to_guests']
      @active_token_value = attributes['active_token_value']
      @current_users = attributes['users'].map do |user|
        user[:created_at] = Time.parse(user[:created_at])
        user
      end

      @loaded = true
    end

    def send_message(message, type = 'TextMessage')
      post 'speak', {:message => {:body => message, :type => type}}
    end

    def get(action)
      connection.get(room_url_for(action))
    end

    def post(action, body = nil)
      connection.post(room_url_for(action), body)
    end

    def raw_post(action, body = nil)
      connection.raw_post(room_url_for(action), body)
    end

    def room_url_for(action, format="json")
      "/room/#{@id}/#{action}.#{format}"
    end

    def connection
      @connection
    end

  end
end
